soco/zgs.py

```diff
@@ -0,0 +1,267 @@
+            _LOG.debug("Setting ZGS cache to %ss", timeout)
+
+        tree = normalize_zgs_xml(payload)
+        normalized_zgs = str(tree)
+        if normalized_zgs == self._last_zgs:
+            _LOG.debug(
+                "Duplicate ZGS received from %s (%s), ignoring", source_ip, source
+            )
+            update_cache()
+            return
+
+        self.processed_count += 1
+        _LOG.debug(
+            "Updating ZGS with %s payload from %s (%s/%s processed)",
+            source,
+            source_ip,
+            self.processed_count,
+            self.total_requests,
+        )
+
+        self.update_soco_instances(tree)
+        update_cache()
+        self._last_zgs = normalized_zgs
+
+    def parse_zone_group_member(self, member_element):
+        """Parse a ZoneGroupMember or Satellite element from Zone Group
+        State, create a SoCo instance for the member, set basic attributes
+        and return it."""
+        # pylint: disable=protected-access
+
+        # Create a SoCo instance for each member. Because SoCo
+        # instances are singletons, this is cheap if they have already
+        # been created, and useful if they haven't. We can then
+        # update various properties for that instance.
+        member_attribs = member_element.attrib
+        ip_addr = member_attribs["Location"].split("//")[1].split(":")[0]
+        zone = config.SOCO_CLASS(ip_addr)
+        for key, attrib in ZGS_ATTRIB_MAPPING.items():
+            setattr(zone, attrib, member_attribs.get(key))
+
+        for channel_map in list(
+            filter(None, [zone._channel_map, zone._ht_sat_chan_map])
+        ):
+            for channel in channel_map.split(";"):
+                if channel.startswith(zone._uid):
+                    zone._channel = channel.split(":")[-1]
+
+        # Add the zone to the set of all members, and to the set
+        # of visible members if appropriate
+        if member_attribs.get("Invisible") != "1":
+            self.visible_zones.add(zone)
+        self.all_zones.add(zone)
+        return zone
+
+    def update_soco_instances(self, tree):
+        """Update all SoCo instances with the provided payload."""
+        # pylint: disable=protected-access
+        self.clear_zone_groups()
+
+        # Compatibility fallback for pre-10.1 firmwares
+        # where a "ZoneGroups" element is not used
+        zone_groups = tree.find("ZoneGroups")
+        if zone_groups is None:
+            zone_groups = tree
+
+        for group_element in zone_groups.findall("ZoneGroup"):
+            coordinator_uid = group_element.attrib["Coordinator"]
+            group_uid = group_element.attrib["ID"]
+            group_coordinator = None
+            members = set()
+            for member_element in group_element.findall("ZoneGroupMember"):
+                zone = self.parse_zone_group_member(member_element)
+                zone._is_satellite = False
+                zone._satellite_parent = None
```

Comment:
It would be cleaner if you would set these inside `parse_zone_group_member`. Same applies probably on all other cases below, where a protected member is being set based on accesses on `member_element`.

Reply:
It could, but `parse_zone_group_member` is called for both the primary & satellite devices in this method. In the latter case it would be set these values to the defaults and then immediately override them after the call returns. I think the current logic is _slightly_ easier to follow as it sets both to the expected value once.